test/framework/client/collect.go

```diff
@@ -389,6 +397,95 @@ func CollectFailure(cluster framework.Cluster, container, destination string, fn
 	return response, nil
 }
 
+func CollectResponsesAndFailures(
+	cluster framework.Cluster,
+	container, destination string,
+	fn ...CollectResponsesOptsFn,
+) ([]FailureResponse, error) {
+	opts := CollectOptions(destination, fn...)
+	cmd := collectCommand(opts, "curl",
+		"--request", opts.Method,
+		"--max-time", "3",
+		"--silent",               // Suppress human-readable errors.
+		"--write-out", "%{json}", // Write JSON result. Requires curl 7.70.0, April 2020.
+		// Silence output so that we don't try to parse it. A future refactor could try to address this
+		// by using "%{stderr}%{json}", but that needs a bit more investigation.
+		"--output", os.DevNull,
+		opts.ShellEscaped(opts.URL),
+	)
+
+	var appPodName string
+	if opts.namespace != "" && opts.application != "" {
+		var err error
+		appPodName, err = framework.PodNameOfApp(cluster, opts.application, opts.namespace)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	mut := sync.Mutex{}
+	var responses []FailureResponse
+
+	var wg sync.WaitGroup
+	var err error
+	for i := 0; i < opts.NumberOfRequests; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			stdout, _, localErr := cluster.Exec(opts.namespace, appPodName, container, cmd...)
+
+			// 1. If we fail to decode the JSON status, return the JSON error,
+			// but prefer the original error if we have it.
+			empty := FailureResponse{}
+			response := FailureResponse{}
+			if jsonErr := json.Unmarshal([]byte(stdout), &response); jsonErr != nil {
+				// Prefer the original error to a JSON decoding error.
+				if localErr == nil {
+					err = jsonErr
+					return
+				}
+			}
+
+			// 2. If there was no error response, we still prefer the original
+			// error, but fall back to reporting that the JSON  is missing.
+			if response == empty {
+				if localErr != nil {
+					err = localErr
+					return
+				}
+
+				err = errors.Errorf("empty JSON response from curl: %q", stdout)
+				return
+			}
+
+			// for k8s
+			k8sExitErr := k8s_exec.CodeExitError{}
+			if errors.As(err, &k8sExitErr) {
+				response.Exitcode = k8sExitErr.Code
+			}
+			// for universal
+			var exitErr *exec.ExitError
+			if errors.As(err, &exitErr) {
+				response.Exitcode = exitErr.ExitCode()
+			}
+
+			mut.Lock()
+			responses = append(responses, response)
+			mut.Unlock()
+		}()
+	}
+	wg.Wait()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. Finally, report the JSON status and no execution error
+	// since the JSON contains all the Curl error information.
+	return responses, nil
+}
```

Comment:
Is it not possible to have something like
```
for i := 0; i < numofrequests; i++ {
  go func() {
    CollectFailure
  }()
}
```
?

Reply:
addressed